src/dotty/tools/dotc/transform/ClassOf.scala

```diff
@@ -0,0 +1,49 @@
+package dotty.tools.dotc
+package transform
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Decorators.StringDecorator
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.TypeErasure
+import dotty.tools.dotc.transform.TreeTransforms.{MiniPhaseTransform, TransformerInfo, TreeTransform}
+
+class ClassOf extends MiniPhaseTransform {
+  import tpd._
+
+  override def phaseName: String = "classOf"
+
+  private var classOfMethod: TermSymbol = _
+
+  override def prepareForUnit(tree: tpd.Tree)(implicit ctx: Context): TreeTransform = {
+    val predefModule = ctx.definitions.ScalaPredefModule
+    classOfMethod = ctx.requiredMethod(predefModule.moduleClass.asClass, "classOf".toTermName)
+    this
+  }
+
+  override def transformTypeApply(tree: TypeApply)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    if (tree.symbol eq classOfMethod) {
+      val tp = tree.args.head.tpe
+      val defn = ctx.definitions
+      val claz = tp.classSymbol
+      if (ValueClasses.isDerivedValueClass(claz)) {
+        Literal(Constant(ref(claz).tpe))
+      } else {
+        def TYPE(module: TermSymbol) = ref(module).select("TYPE".toTermName).ensureConforms(tree.tpe)
+        claz match {
+          case defn.BooleanClass => TYPE(defn.BoxedBooleanModule)
+          case defn.ByteClass    => TYPE(defn.BoxedByteModule)
+          case defn.ShortClass   => TYPE(defn.BoxedShortModule)
+          case defn.CharClass    => TYPE(defn.BoxedCharModule)
+          case defn.IntClass     => TYPE(defn.BoxedIntModule)
+          case defn.LongClass    => TYPE(defn.BoxedLongModule)
+          case defn.FloatClass   => TYPE(defn.BoxedFloatModule)
+          case defn.DoubleClass  => TYPE(defn.BoxedDoubleModule)
+          case defn.UnitClass    => TYPE(defn.BoxedVoidModule)
+          case _                 => Literal(Constant(TypeErasure.erasure(tp)))
```

Comment:
This doesn't work. The output is as follows:
```scala
println(classOf[Unit]) // class scala.runtime.BoxedUnit instead of void
println(classOf[Boolean]) // class java.lang.Boolean instead of boolean
println(classOf[Byte]) // class java.lang.Byte instead of byte
println(classOf[Short]) // class java.lang.Short instead of short
println(classOf[Char]) // class java.lang.Character instead of char
println(classOf[Int]) // class java.lang.Integer instead of int
println(classOf[Long]) // class java.lang.Long instead of long
println(classOf[Float]) // class java.lang.Float instead of float
println(classOf[Double]) // class java.lang.Double instead of double
```